File: python/examples/01_the_switch_and_the_light/light.py
import atexit
import ssl
import sys
import time
import RPi.GPIO as GPIO

sys.path.append('../../src/arete_sdk')
from client import Client

GPIO23 = 23

NODE_ID = 'onqXVczGoymQkFc3UN6qcM'
NODE_NAME = 'arete-sdk-01-light'

CONTEXT_ID = 'uRLoYsXEY7nsbs9fRdjM8A'
CONTEXT_NAME = 'Building 23, Office 41-B'

PADI_LIGHT_PROFILE = 'padi.light'

# Configure pin for output
GPIO.setmode(GPIO.BCM)
GPIO.setup(GPIO23, GPIO.OUT)

# Connect to Arete control plane
ssl_context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_CLIENT)
ssl_context.check_hostname = False
ssl_context.verify_mode = ssl.CERT_NONE
client = Client.connect('wss://dashboard.test.cns.dev:443', ssl=ssl_context)
client.wait_for_open()
sys.stderr.write('Connected to Arete control plane\n')

# Register with the control plane
system = client.system()
node = system.node(NODE_ID, NODE_NAME)
context = node.context(CONTEXT_ID, CONTEXT_NAME)


# Detect initial desired state, plus future changes to desired state, and try to actualize it
def detect_change(event):
    if event['property'] == 'sOut':
        desired_state = event['value'] == '1'
        GPIO.output(GPIO23, desired_state)
        sys.stderr.write('Light is now {}\n'.format('ON' if desired_state else 'OFF'))


consumer = context.consumer(PADI_LIGHT_PROFILE)
consumer.watch(detect_change)


# Register shutdown handling
@atexit.register
def cleanup():
    GPIO.cleanup()


# Startup complete
sys.stderr.write('Light service started\n')
while True:
    time.sleep(0.1)

File: nodejs/examples/01_the_switch_and_the_light/light.js
import { Gpio } from 'onoff';
import { Client } from '../../index.js';

const GPIO23 = 535;

const CONTEXT_ID = 'uRLoYsXEY7nsbs9fRdjM8A';
const CONTEXT_NAME = 'Building 23, Office 41-B';

const NODE_ID = 'onqXVczGoymQkFc3UN6qcM';
const NODE_NAME = 'arete-sdk-01-light';

const PADI_LIGHT_PROFILE = 'padi.light';

// Connect to Arete control plane
let client = new Client({
  protocol: 'wss:',
  host: 'dashboard.test.cns.dev',
  port: 443,
});
await client.waitForOpen(5000);
console.log('Connected to Arete control plane');

// Register with the control plane
let system = await client.system();
let node = await system.node(NODE_ID, NODE_NAME, false);
let context = await node.context(CONTEXT_ID, CONTEXT_NAME);

// Read initial actual state of the light, and sync it with Arete
let pin = new Gpio(GPIO23, 'in', 'both', { debounceTimeout: 10 });
let state = pin.readSync();
let consumer = await context.consumer(PADI_LIGHT_PROFILE);
consumer.put('cState', state ? '1' : '0');
console.log('Light is initially', state ? 'ON' : 'OFF');

// Detect initial desired state, plus future changes to desired state, and try to actualize it
consumer.watch((event) => {
  if (event.property == 'sOut') {
    const desiredState = event.value == '1';
    let pin = new Gpio(GPIO23, 'out');
    pin.writeSync(desiredState ? Gpio.HIGH : Gpio.LOW, (err) => {
      if (err) {
        throw err;
      }
    });
    consumer.put('cState', desiredState ? '1' : '0');
    console.log('Light is now', desiredState ? 'ON' : 'OFF');
  }
});

// Register shutdown handling
process.on('SIGINT', (_) => {
  console.log();
  console.log('Light service terminating');
  pin.unexport();
});

// Startup complete
console.log('Light service started');
